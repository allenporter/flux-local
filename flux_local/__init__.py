"""
.. include:: ../README.md
"""

__all__ = [
    "git_repo",
    "manifest",
    "kustomize",
    "helm",
    "exceptions",
]
