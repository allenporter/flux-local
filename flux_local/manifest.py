"""Representation of the contents of a cluster.

A manifest may be built directly from the local context of a cluster, or may be
serialized and stored and checked into the cluster for use in other applications
e.g. such as writing management plan for resources.
"""

from pathlib import Path
from typing import Any, Optional, cast

import aiofiles
import yaml

try:
    from pydantic.v1 import BaseModel, Field
except ImportError:
    from pydantic import BaseModel, Field  # type: ignore

from .exceptions import InputException

__all__ = [
    "read_manifest",
    "write_manifest",
    "Manifest",
    "Cluster",
    "Kustomization",
    "HelmRepository",
    "HelmRelease",
    "HelmChart",
    "ClusterPolicy",
]

# Match a prefix of apiVersion to ensure we have the right type of object.
# We don't check specific versions for forward compatibility on upgrade.
FLUXTOMIZE_DOMAIN = "kustomize.toolkit.fluxcd.io"
KUSTOMIZE_DOMAIN = "kustomize.config.k8s.io"
HELM_REPO_DOMAIN = "source.toolkit.fluxcd.io"
HELM_RELEASE_DOMAIN = "helm.toolkit.fluxcd.io"
CLUSTER_POLICY_DOMAIN = "kyverno.io"
CRD_KIND = "CustomResourceDefinition"
SECRET_KIND = "Secret"

REPO_TYPE_DEFAULT = "default"
REPO_TYPE_OCI = "oci"


def _check_version(doc: dict[str, Any], version: str) -> None:
    """Assert that the resource has the specified version."""
    if not (api_version := doc.get("apiVersion")):
        raise InputException(f"Invalid object missing apiVersion: {doc}")
    if not api_version.startswith(version):
        raise InputException(f"Invalid object expected '{version}': {doc}")


class BaseManifest(BaseModel):
    """Base class for all manifest objects."""

    _COMPACT_EXCLUDE_FIELDS: dict[str, Any] = {}

    def compact_dict(self, exclude: dict[str, Any] | None = None) -> dict[str, Any]:
        """Return a compact dictionary representation of the object.

        This is similar to `dict()` but with a specific implementation for serializing
        with variable fields removed.
        """
        if exclude is None:
            exclude = self._COMPACT_EXCLUDE_FIELDS
        return self.dict(exclude=exclude)  # type: ignore[arg-type]

    @classmethod
    def parse_yaml(cls, content: str) -> "BaseManifest":
        """Parse a serialized manifest."""
        doc = next(yaml.load_all(content, Loader=yaml.Loader), None)
        return cls.parse_obj(doc)

    def yaml(self, exclude: dict[str, Any] | None = None) -> str:
        """Return a YAML string representation of compact_dict."""
        data = self.compact_dict(exclude)
        return yaml.dump(data, sort_keys=False, explicit_start=True)


class HelmChart(BaseManifest):
    """A representation of an instantiation of a chart for a HelmRelease."""

    name: str
    """The name of the chart within the HelmRepository."""

    version: Optional[str] = None
    """The version of the chart."""

    repo_name: str
    """The name of the HelmRepository."""

    repo_namespace: str
    """The namespace of the HelmRepository."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any], default_namespace: str) -> "HelmChart":
        """Parse a HelmChart from a HelmRelease resource object."""
        _check_version(doc, HELM_RELEASE_DOMAIN)
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        if not (chart := spec.get("chart")):
            raise InputException(f"Invalid {cls} missing spec.chart: {doc}")
        if not (chart_spec := chart.get("spec")):
            raise InputException(f"Invalid {cls} missing spec.chart.spec: {doc}")
        if not (chart := chart_spec.get("chart")):
            raise InputException(f"Invalid {cls} missing spec.chart.spec.chart: {doc}")
        version = chart_spec.get("version")
        if not (source_ref := chart_spec.get("sourceRef")):
            raise InputException(
                f"Invalid {cls} missing spec.chart.spec.sourceRef: {doc}"
            )
        if "name" not in source_ref:
            raise InputException(f"Invalid {cls} missing sourceRef fields: {doc}")
        return cls(
            name=chart,
            version=version,
            repo_name=source_ref["name"],
            repo_namespace=source_ref.get("namespace", default_namespace),
        )

    @property
    def chart_name(self) -> str:
        """Identifier for the HelmChart."""
        return f"{self.repo_namespace}-{self.repo_name}/{self.name}"

    _COMPACT_EXCLUDE_FIELDS = {"version": True}


class HelmRelease(BaseManifest):
    """A representation of a Flux HelmRelease."""

    name: str
    """The name of the HelmRelease."""

    namespace: str
    """The namespace that owns the HelmRelease."""

    chart: HelmChart
    """A mapping to a specific helm chart for this HelmRelease."""

    values: Optional[dict[str, Any]] = None
    """The values to install in the chart."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "HelmRelease":
        """Parse a HelmRelease from a kubernetes resource object."""
        _check_version(doc, HELM_RELEASE_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        chart = HelmChart.parse_doc(doc, namespace)
        return cls(
            name=name,
            namespace=namespace,
            chart=chart,
            values=doc["spec"].get("values"),
        )

    @property
    def release_name(self) -> str:
        """Identifier for the HelmRelease."""
        return f"{self.namespace}-{self.name}"

    @property
    def repo_name(self) -> str:
        """Identifier for the HelmRepository identified in the HelmChart."""
        return f"{self.chart.repo_namespace}-{self.chart.repo_name}"

    _COMPACT_EXCLUDE_FIELDS = {
        "values": True,
        "chart": HelmChart._COMPACT_EXCLUDE_FIELDS,
    }


class HelmRepository(BaseManifest):
    """A representation of a flux HelmRepository."""

    name: str
    """The name of the HelmRepository."""

    namespace: str
    """The namespace of owning the HelmRepository."""

    url: str
    """The URL to the repository of helm charts."""

    repo_type: str | None = None
    """The type of the HelmRepository."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "HelmRepository":
        """Parse a HelmRepository from a kubernetes resource."""
        _check_version(doc, HELM_REPO_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        if not (url := spec.get("url")):
            raise InputException(f"Invalid {cls} missing spec.url: {doc}")
        return cls(
            name=name,
            namespace=namespace,
            url=url,
            repo_type=spec.get("type", REPO_TYPE_DEFAULT),
        )

    @property
    def repo_name(self) -> str:
        """Identifier for the HelmRepository."""
        return f"{self.namespace}-{self.name}"


class ClusterPolicy(BaseManifest):
    """A kyverno policy object."""

    name: str
    """The name of the kustomization."""

    namespace: str | None = None
    """The namespace of the kustomization."""

    doc: dict[str, Any] | None = None
    """The raw ClusterPolicy document."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "ClusterPolicy":
        """Parse a cluster policy object from a kubernetes resource."""
        _check_version(doc, CLUSTER_POLICY_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        namespace = metadata.get("namespace")
        if not doc.get("spec"):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        return ClusterPolicy(name=name, namespace=namespace, doc=doc)

    _COMPACT_EXCLUDE_FIELDS = {
        "doc": True,
    }


class Kustomization(BaseManifest):
    """A Kustomization is a set of declared cluster artifacts.

    This represents a flux Kustomization that points to a path that
    contains typical `kustomize` Kustomizations on local disk that
    may be flat or contain overlays.
    """

    name: str
    """The name of the kustomization."""

    namespace: str | None = None
    """The namespace of the kustomization."""

    path: str
    """The local repo path to the kustomization contents."""

    helm_repos: list[HelmRepository] = Field(default_factory=list)
    """The set of HelmRepositories represented in this kustomization."""

    helm_releases: list[HelmRelease] = Field(default_factory=list)
    """The set of HelmRelease represented in this kustomization."""

    cluster_policies: list[ClusterPolicy] = Field(default_factory=list)
    """The set of ClusterPolicies represented in this kustomization."""

    source_path: str | None = None
    """Optional source path for this Kustomization, relative to the build path."""

    source_kind: str | None = None
    """The sourceRef kind that provides this Kustomization e.g. GitRepository etc."""

    source_name: str | None = None
    """The name of the sourceRef that provides this Kustomization."""

    source_namespace: str | None = None
    """The namespace of the sourceRef that provides this Kustomization."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "Kustomization":
        """Parse a partial Kustomization from a kubernetes resource."""
        _check_version(doc, FLUXTOMIZE_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        path = spec.get("path", "")
        source_path = metadata.get("annotations", {}).get("config.kubernetes.io/path")
        source_ref = spec.get("sourceRef", {})
        return Kustomization(
            name=name,
            namespace=namespace,
            path=path,
            source_path=source_path,
            source_kind=source_ref.get("kind"),
            source_name=source_ref.get("name"),
            source_namespace=source_ref.get("namespace", namespace),
        )

    @property
    def id_name(self) -> str:
        """Identifier for the Kustomization in tests"""
        return f"{self.path}"

    @property
    def namespaced_name(self, sep: str = "/") -> str:
        """Return the namespace and name concatenated as an id."""
        return f"{self.namespace}{sep}{self.name}"

    _COMPACT_EXCLUDE_FIELDS = {
        "helm_releases": {
            "__all__": HelmRelease._COMPACT_EXCLUDE_FIELDS,
        },
        "cluster_policies": {
            "__all__": ClusterPolicy._COMPACT_EXCLUDE_FIELDS,
        },
        "source_path": True,
        "source_name": True,
        "source_namespace": True,
        "source_kind": True,
    }


class Cluster(BaseManifest):
    """A set of nodes that run containerized applications.

    Many flux git repos will only have a single flux cluster, though
    a repo may also contain multiple (e.g. dev an prod).
    """

    name: str
    """The name of the cluster."""

    namespace: str
    """The namespace of the cluster."""

    path: str
    """The local git repo path to the Kustomization objects for the cluster."""

    kustomizations: list[Kustomization] = Field(default_factory=list)
    """A list of flux Kustomizations for the cluster."""

    @classmethod
    def parse_doc(cls, doc: dict[str, Any]) -> "Cluster":
        """Parse a partial Kustomization from a kubernetes resource."""
        _check_version(doc, FLUXTOMIZE_DOMAIN)
        if not (metadata := doc.get("metadata")):
            raise InputException(f"Invalid {cls} missing metadata: {doc}")
        if not (name := metadata.get("name")):
            raise InputException(f"Invalid {cls} missing metadata.name: {doc}")
        if not (namespace := metadata.get("namespace")):
            raise InputException(f"Invalid {cls} missing metadata.namespace: {doc}")
        if not (spec := doc.get("spec")):
            raise InputException(f"Invalid {cls} missing spec: {doc}")
        if not (path := spec.get("path")):
            raise InputException(f"Invalid {cls} missing spec.path: {doc}")
        return Cluster(name=name, namespace=namespace, path=path)

    @property
    def namespaced_name(self, sep: str = "/") -> str:
        """Return the namespace and name concatenated as an id."""
        return f"{self.namespace}{sep}{self.name}"

    @property
    def id_name(self) -> str:
        """Identifier for the Cluster in tests."""
        return f"{self.path}"

    @property
    def helm_repos(self) -> list[HelmRepository]:
        """Return the list of HelmRepository objects from all Kustomizations."""
        return [
            repo
            for kustomization in self.kustomizations
            for repo in kustomization.helm_repos
        ]

    @property
    def helm_releases(self) -> list[HelmRelease]:
        """Return the list of HelmRelease objects from all Kustomizations."""
        return [
            release
            for kustomization in self.kustomizations
            for release in kustomization.helm_releases
        ]

    @property
    def cluster_policies(self) -> list[ClusterPolicy]:
        """Return the list of ClusterPolicy objects from all Kustomizations."""
        return [
            policy
            for kustomization in self.kustomizations
            for policy in kustomization.cluster_policies
        ]

    _COMPACT_EXCLUDE_FIELDS = {
        "kustomizations": {
            "__all__": Kustomization._COMPACT_EXCLUDE_FIELDS,
        }
    }


class Manifest(BaseManifest):
    """Holds information about cluster and applications contained in a repo."""

    clusters: list[Cluster]
    """A list of Clusters represented in the repo."""

    _COMPACT_EXCLUDE_FIELDS = {
        "clusters": {
            "__all__": Cluster._COMPACT_EXCLUDE_FIELDS,
        }
    }


async def read_manifest(manifest_path: Path) -> Manifest:
    """Return the contents of a serialized manifest file.

    A manifest file is typically created by `flux-local get cluster -o yaml` or
    similar command.
    """
    async with aiofiles.open(str(manifest_path)) as manifest_file:
        content = await manifest_file.read()
        if not content:
            raise ValueError("validation error for Manifest file {manifest_path}")
        return cast(Manifest, Manifest.parse_yaml(content))


async def write_manifest(manifest_path: Path, manifest: Manifest) -> None:
    """Write the specified manifest content to disk."""
    content = manifest.yaml()
    async with aiofiles.open(str(manifest_path), mode="w") as manifest_file:
        await manifest_file.write(content)


async def update_manifest(manifest_path: Path, manifest: Manifest) -> None:
    """Write the specified manifest only if changed."""
    new_content = manifest.yaml()
    async with aiofiles.open(str(manifest_path)) as manifest_file:
        content = await manifest_file.read()
    if content == new_content:
        return
    async with aiofiles.open(str(manifest_path), mode="w") as manifest_file:
        await manifest_file.write(new_content)
