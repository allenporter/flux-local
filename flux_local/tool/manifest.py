"""Flux-local manifest action."""

import pathlib

from flux_local import git_repo


class ManifestAction:
    """Flux-local manifest action."""

    async def run(  # type: ignore[no-untyped-def]
        self,
        path: pathlib.Path,
        **kwargs,  # pylint: disable=unused-argument
    ) -> None:
        """Async Action implementation."""
        manifest = await git_repo.build_manifest(path)
        print(manifest.yaml())
