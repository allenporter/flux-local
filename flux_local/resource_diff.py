"""Module for computing resource diffs.

This is used internally, primarily by the diff tool.
"""

from collections.abc import Iterable
from dataclasses import asdict
import difflib
import logging
import pathlib
import tempfile
from typing import Generator, Any, AsyncGenerator, TypeVar
import yaml


from . import command
from .visitor import ObjectOutput, ResourceKey, HelmVisitor
from .manifest import HelmRelease, NamedResource

_LOGGER = logging.getLogger(__name__)

_TRUNCATE = "[Diff truncated by flux-local]"

T = TypeVar("T")


def _unique_keys(k1: dict[T, Any], k2: dict[T, Any]) -> Iterable[T]:
    """Return an ordered set."""
    return {
        **{k: True for k in k1.keys()},
        **{k: True for k in k2.keys()},
    }.keys()


def perform_object_diff(
    a: ObjectOutput, b: ObjectOutput, n: int, limit_bytes: int
) -> Generator[str, None, None]:
    """Generate diffs between the two output objects."""
    for kustomization_key in _unique_keys(a.content, b.content):
        _LOGGER.debug(
            "Diffing results for Kustomization %s (n=%d)", kustomization_key, n
        )
        a_resources = a.content.get(kustomization_key, {})
        b_resources = b.content.get(kustomization_key, {})
        for resource_key in _unique_keys(a_resources, b_resources):
            diff_text = difflib.unified_diff(
                a=a_resources.get(resource_key, []),
                b=b_resources.get(resource_key, []),
                fromfile=f"{kustomization_key.label} {resource_key.compact_label}",
                tofile=f"{kustomization_key.label} {resource_key.compact_label}",
                n=n,
            )
            size = 0
            for line in diff_text:
                size += len(line)
                if limit_bytes and size > limit_bytes:
                    yield _TRUNCATE
                    break
                yield line


async def perform_external_diff(
    cmd: list[str],
    a: ObjectOutput,
    b: ObjectOutput,
    limit_bytes: int,
) -> AsyncGenerator[str, None]:
    """Generate diffs between the two output objects."""
    with tempfile.TemporaryDirectory() as tmpdir:
        for kustomization_key in _unique_keys(a.content, b.content):
            _LOGGER.debug(
                "Diffing results for Kustomization %s",
                kustomization_key,
            )
            a_resources = a.content.get(kustomization_key, {})
            b_resources = b.content.get(kustomization_key, {})
            keys = _unique_keys(a_resources, b_resources)

            a_file = pathlib.Path(tmpdir) / "a.yaml"
            a_file.write_text(
                "\n".join(
                    [
                        "\n".join(a_resources.get(resource_key, []))
                        for resource_key in keys
                    ]
                )
            )
            b_file = pathlib.Path(tmpdir) / "b.yaml"
            b_file.write_text(
                "\n".join(
                    [
                        "\n".join(b_resources.get(resource_key, []))
                        for resource_key in keys
                    ]
                )
            )

            out = await command.Command(
                cmd + [str(a_file), str(b_file)], retcodes=[0, 1]
            ).run()
            if out:
                result = out.decode("utf-8")
                if limit_bytes and len(result) > limit_bytes:
                    result = result[:limit_bytes] + "\n" + _TRUNCATE
                yield result


def _omit_none(obj: Any) -> dict[str, Any]:
    """Creates a dictionary with None values missing."""
    return {k: v for k, v in obj if v is not None}


def perform_yaml_diff(
    a: ObjectOutput,
    b: ObjectOutput,
    n: int,
    limit_bytes: int,
) -> Generator[str, None, None]:
    """Generate diffs between the two output objects."""

    diffs = []
    for kustomization_key in _unique_keys(a.content, b.content):
        _LOGGER.debug("Diffing results for %s (n=%d)", kustomization_key, n)
        a_resources = a.content.get(kustomization_key, {})
        b_resources = b.content.get(kustomization_key, {})
        resource_diffs = []
        for resource_key in _unique_keys(a_resources, b_resources):
            diff_text = difflib.unified_diff(
                a=a_resources.get(resource_key, []),
                b=b_resources.get(resource_key, []),
                fromfile=f"{kustomization_key.label} {resource_key.compact_label}",
                tofile=f"{kustomization_key.label} {resource_key.compact_label}",
                n=n,
            )
            diff_content = "\n".join(diff_text)
            if not diff_content:
                continue
            if limit_bytes and len(diff_content) > limit_bytes:
                diff_content = diff_content[:limit_bytes] + "\n" + _TRUNCATE
            obj = {
                **asdict(resource_key, dict_factory=_omit_none),
                "diff": diff_content,
            }
            resource_diffs.append(obj)
        if resource_diffs:
            diffs.append(
                {
                    **asdict(kustomization_key),
                    "diffs": resource_diffs,
                }
            )
    if diffs:
        yield yaml.dump(diffs, sort_keys=False, explicit_start=True, default_style=None)


def merge_helm_releases(
    a: list[HelmRelease], b: list[HelmRelease]
) -> Generator[tuple[HelmRelease | None, HelmRelease | None], None, None]:
    """Return HelmReleases joined by name."""
    a_dict = {r.release_name: r for r in a}
    b_dict = {r.release_name: r for r in b}
    for name in _unique_keys(a_dict, b_dict):
        yield a_dict.get(name), b_dict.get(name)


def merge_named_resources(
    a: list[NamedResource], b: list[NamedResource]
) -> Generator[NamedResource, None, None]:
    """Return HelmReleases joined by name."""
    a_dict = {f"{r.kind}-{r.namespaced_name}": r for r in a}
    b_dict = {f"{r.kind}-{r.namespaced_name}": r for r in b}
    for name in _unique_keys(a_dict, b_dict):
        # Emit either the left or right since they should be identical
        value: NamedResource = a_dict.get(name) or b_dict.get(name)  # type: ignore[assignment]
        yield value


def build_helm_dependency_map(
    a_visitor: HelmVisitor, b_visitor: HelmVisitor
) -> dict[NamedResource, NamedResource]:
    """Return a map of all resources to the HelmRelease resource that depends on them.

    This is a mapping of all the named resources. If any of them changed, then we
    need to diff the HelmRelease to look for changes in the rendered output.
    """
    results: dict[NamedResource, NamedResource] = {}
    for helmrelease_a, helmrelease_b in merge_helm_releases(
        a_visitor.releases, b_visitor.releases
    ):
        resources_a = helmrelease_a.resource_dependencies if helmrelease_a else []
        resources_b = helmrelease_b.resource_dependencies if helmrelease_b else []
        hr: HelmRelease = helmrelease_a or helmrelease_b  # type: ignore[assignment]
        hr_named_resource = NamedResource(
            kind="HelmRelease",
            namespace=hr.namespace,
            name=hr.name,
        )
        for named_resource in merge_named_resources(resources_a, resources_b):
            results[named_resource] = hr_named_resource

    return results


def get_helm_release_diff_keys(
    a: ObjectOutput, b: ObjectOutput, dependency_map: dict[NamedResource, NamedResource]
) -> list[ResourceKey]:
    """Return HelmRelease resource keys with diffs, by cluster."""

    # Pass #1: Check all named resources that are upstream dependencies of a
    # HelmRelease and have a content change. Save the HelmRelease name.
    hrs_to_check: set[NamedResource] = set()
    _LOGGER.debug("Checking HelmRelease dependencies")
    for kustomization_key in _unique_keys(a.content, b.content):
        _LOGGER.debug("Diffing results for Kustomization %s", kustomization_key)
        a_resources = a.content.get(kustomization_key, {})
        b_resources = b.content.get(kustomization_key, {})
        for resource_key in _unique_keys(a_resources, b_resources):
            if (
                hr_named_resource := dependency_map.get(resource_key.named_resource)
            ) is None:
                continue
            if a_resources.get(resource_key) != b_resources.get(resource_key):
                _LOGGER.info(
                    "HelmRelease %s detected potential change in resource %s (Kustomization %s)",
                    hr_named_resource.namespaced_name,
                    resource_key.namespaced_name,
                    kustomization_key.namespaced_name,
                )
                hrs_to_check.add(hr_named_resource)

    # Pass #2: Find the ResourceKey of all the changed HelmReleases
    results: list[ResourceKey] = []
    for kustomization_key in _unique_keys(a.content, b.content):
        a_resources = a.content.get(kustomization_key, {})
        b_resources = b.content.get(kustomization_key, {})
        for resource_key in _unique_keys(a_resources, b_resources):
            if resource_key.named_resource in hrs_to_check:
                results.append(resource_key)

    return results
