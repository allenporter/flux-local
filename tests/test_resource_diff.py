"""Tests for the resource diff module."""

import pathlib
import shutil
import tempfile
from contextlib import contextmanager
from collections.abc import Generator

import yaml

from flux_local import git_repo
from flux_local.resource_diff import (
    get_helm_release_diff_keys,
    merge_helm_releases,
    merge_named_resources,
    build_helm_dependency_map,
)
from flux_local.visitor import ObjectOutput, HelmVisitor, ResourceKey
from flux_local.manifest import HelmRelease, HelmChart, NamedResource

TESTDATA_PATH = pathlib.Path("tests/testdata/cluster8")


async def visit_helm_content(path: pathlib.Path) -> tuple[ObjectOutput, HelmVisitor]:
    """Visit content in the specified path."""

    query = git_repo.ResourceSelector()
    query.path = git_repo.PathSelector(path)
    query.helm_release.namespace = None  # All

    content = ObjectOutput([])
    query.kustomization.visitor = content.visitor()

    helm_visitor = HelmVisitor()
    query.helm_repo.visitor = helm_visitor.repo_visitor()
    query.helm_release.visitor = helm_visitor.release_visitor()

    await git_repo.build_manifest(selector=query)

    return content, helm_visitor


@contextmanager
def mirror_worktree(path: pathlib.Path) -> Generator[pathlib.Path]:
    """Create a copy of the path in a temporary git repo for generating diffs."""
    with tempfile.TemporaryDirectory() as tmpdir:
        tmp_path = pathlib.Path(tmpdir)
        # Required by `git_repo` to find the root
        shutil.copytree(".git", str(tmp_path / ".git"))

        new_path = tmp_path / path
        shutil.copytree(path, new_path)

        yield new_path


def test_no_diff_keys_empty_input() -> None:
    """Test get_helm_release_diff_keys with empty inputs."""

    a = ObjectOutput([])
    b = ObjectOutput([])
    assert get_helm_release_diff_keys(a, b, {}) == []


async def test_no_diff_keys() -> None:
    """Test get_helm_release_diff_keys with no diff."""

    content, helm_visitor = await visit_helm_content(TESTDATA_PATH)
    with mirror_worktree(TESTDATA_PATH) as new_cluster_path:
        content_new, helm_visitor_new = await visit_helm_content(new_cluster_path)

    dep_map = build_helm_dependency_map(helm_visitor, helm_visitor_new)
    assert get_helm_release_diff_keys(content, content_new, dep_map) == []


async def test_helm_release_spec_diff() -> None:
    """Test get_helm_release_diff_keys with a diff in the input values."""

    content, helm_visitor = await visit_helm_content(TESTDATA_PATH)
    with mirror_worktree(TESTDATA_PATH) as new_cluster_path:

        # Generate a diff by changing a value in the HelmRelease
        podinfo_path = new_cluster_path / "apps/podinfo.yaml"
        podinfo_doc = list(
            yaml.load_all(podinfo_path.read_text(), Loader=yaml.SafeLoader)
        )
        assert len(podinfo_doc) == 2
        assert podinfo_doc[1]["kind"] == "HelmRelease"
        assert podinfo_doc[1]["spec"]["values"]["redis"]["enabled"]
        podinfo_doc[1]["spec"]["values"]["redis"]["enabled"] = False
        podinfo_path.write_text(yaml.dump_all(podinfo_doc))

        content_new, helm_visitor_new = await visit_helm_content(new_cluster_path)

    dep_map = build_helm_dependency_map(helm_visitor, helm_visitor_new)
    assert get_helm_release_diff_keys(content, content_new, dep_map) == [
        ResourceKey(
            kustomization_path=str(TESTDATA_PATH / "apps"),
            kind="HelmRelease",
            namespace="podinfo",
            name="podinfo",
        )
    ]


async def test_helm_repository_causes_diff() -> None:
    """Test get_helm_release_diff_keys where an input HelmRepository change causes a diff."""

    content, helm_visitor = await visit_helm_content(TESTDATA_PATH)
    with mirror_worktree(TESTDATA_PATH) as new_cluster_path:

        # Generate a diff by changing a value in the HelmRelease
        podinfo_path = new_cluster_path / "apps/podinfo.yaml"
        podinfo_doc = list(
            yaml.load_all(podinfo_path.read_text(), Loader=yaml.SafeLoader)
        )
        assert len(podinfo_doc) == 2
        assert podinfo_doc[0]["kind"] == "HelmRepository"
        assert podinfo_doc[0]["spec"]["url"] == "oci://ghcr.io/stefanprodan/charts"
        podinfo_doc[0]["spec"]["url"] = "oci://ghcr.io/stefanprodan/charts2"
        podinfo_path.write_text(yaml.dump_all(podinfo_doc))

        content_new, helm_visitor_new = await visit_helm_content(new_cluster_path)

    dep_map = build_helm_dependency_map(helm_visitor, helm_visitor_new)
    assert get_helm_release_diff_keys(content, content_new, dep_map) == [
        ResourceKey(
            kustomization_path=str(TESTDATA_PATH / "apps"),
            kind="HelmRelease",
            namespace="podinfo",
            name="podinfo",
        )
    ]


async def test_merge_helm_releases() -> None:
    """Test merge_helm_releases."""

    chart = HelmChart(
        name="chart", version="1.0.0", repo_name="example", repo_namespace="flux-system"
    )

    a = [
        HelmRelease(
            name="a", namespace="default", chart=chart, values={"key": "value1"}
        ),
        HelmRelease(
            name="b", namespace="default", chart=chart, values={"key": "value2"}
        ),
    ]
    b = [
        HelmRelease(
            name="b", namespace="default", chart=chart, values={"key": "value3"}
        ),
        HelmRelease(
            name="c", namespace="default", chart=chart, values={"key": "value4"}
        ),
    ]
    assert list(merge_helm_releases(a, b)) == [
        (
            HelmRelease(
                name="a", namespace="default", chart=chart, values={"key": "value1"}
            ),
            None,
        ),
        (
            HelmRelease(
                name="b", namespace="default", chart=chart, values={"key": "value2"}
            ),
            HelmRelease(
                name="b", namespace="default", chart=chart, values={"key": "value3"}
            ),
        ),
        (
            None,
            HelmRelease(
                name="c", namespace="default", chart=chart, values={"key": "value4"}
            ),
        ),
    ]


async def test_merge_named_resources() -> None:
    """Test for merging named resources."""
    a = [
        NamedResource(name="podinfo", namespace="podinfo", kind="HelmRelease"),
        NamedResource(name="podinfo", namespace="podinfo", kind="ConfigMap"),
        NamedResource(name="secret", namespace="podinfo", kind="Secret"),
    ]
    b = [
        NamedResource(name="podinfo", namespace="podinfo", kind="HelmRelease"),
        NamedResource(name="podinfo", namespace="podinfo", kind="ConfigMap"),
        NamedResource(name="podinfo", namespace="flux-system", kind="HelmRepository"),
    ]
    assert list(merge_named_resources(a, b)) == [
        NamedResource(name="podinfo", namespace="podinfo", kind="HelmRelease"),
        NamedResource(name="podinfo", namespace="podinfo", kind="ConfigMap"),
        NamedResource(name="secret", namespace="podinfo", kind="Secret"),
        NamedResource(name="podinfo", namespace="flux-system", kind="HelmRepository"),
    ]
