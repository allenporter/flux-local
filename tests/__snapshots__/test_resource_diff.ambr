# serializer version: 1
# name: test_perform_json_diff
  '''
  [
      {
          "kustomization_path": "tests/testdata/cluster8/apps",
          "kind": "Kustomization",
          "namespace": "flux-system",
          "name": "apps",
          "diffs": [
              {
                  "kustomization_path": "tests/testdata/cluster8/apps",
                  "kind": "HelmRepository",
                  "namespace": "flux-system",
                  "name": "podinfo",
                  "diff": "--- tests/testdata/cluster8/apps Kustomization: flux-system/apps HelmRepository: flux-system/podinfo\n\n+++ tests/testdata/cluster8/apps Kustomization: flux-system/apps HelmRepository: flux-system/podinfo\n\n@@ -8,7 +8,7 @@\n\n   name: podinfo\n   namespace: flux-system\n spec:\n   interval: 5m\n   type: oci\n-  url: oci://ghcr.io/stefanprodan/charts\n+  url: oci://ghcr.io/stefanprodan/charts2\n "
              }
          ]
      }
  ]
  '''
# ---
# name: test_perform_yaml_diff
  '''
  ---
  - kustomization_path: tests/testdata/cluster8/apps
    kind: Kustomization
    namespace: flux-system
    name: apps
    diffs:
    - kustomization_path: tests/testdata/cluster8/apps
      kind: HelmRepository
      namespace: flux-system
      name: podinfo
      diff: "--- tests/testdata/cluster8/apps Kustomization: flux-system/apps HelmRepository:\
        \ flux-system/podinfo\n\n+++ tests/testdata/cluster8/apps Kustomization: flux-system/apps\
        \ HelmRepository: flux-system/podinfo\n\n@@ -8,7 +8,7 @@\n\n   name: podinfo\n\
        \   namespace: flux-system\n spec:\n   interval: 5m\n   type: oci\n-  url: oci://ghcr.io/stefanprodan/charts\n\
        +  url: oci://ghcr.io/stefanprodan/charts2\n "
  
  '''
# ---
